### PR TITLE
Handle Http2Streams that error during stream start

### LINKED ISSUE #19360: HTTP2 streams returned to pool in uncomplete state
It is possible for a stream to error and be returned to the pool with pipes that are not fully complete

A thrown error will return a stream in a questionable state here - https://github.com/dotnet/aspnetcore/blob/03a9e8ec1c47d3c27902daf586c3ad30b0c40c31/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs#L976 - will eventually be dequeue and returned to the pool here - https://github.com/dotnet/aspnetcore/blob/03a9e8ec1c47d3c27902daf586c3ad30b0c40c31/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs#L1063

We shouldn't be holding onto streams in a pool when the pipes are not complete (I noticed this when working to reuse the pipes - calling Reset will throw if the reader and write aren't complete).

Either the return should be moved to a place where we know the stream is safe to return to the pool, or we should keep track of whether the stream is in the right state to return.

@davidfowl @jkotalik @halter73

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspnetcore/issues/19360

Related to issues found in https://github.com/dotnet/aspnetcore/pull/19356

### CODE REVIEW

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2ConnectionTests.cs [hunk 1]
```diff
@@ -24,6 +24,163 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Tests
...
+
+            // Ping will trigger the stream to be returned to the pool so we can assert it
+            await SendPingAsync(Http2PingFrameFlags.NONE);
+            await ExpectAsync(Http2FrameType.PING,
+                withLength: 8,
+                withFlags: (byte)Http2PingFrameFlags.ACK,
+                withStreamId: 0);
+
+            // Stream returned to the pool
+            Assert.Equal(1, _connection.StreamPool.Count);
```
Comment (JamesNK) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2ConnectionTests.cs]: The pooled stream has uncompleted pipes

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs [hunk 2]
```diff
@@ -1060,7 +1066,12 @@ private void UpdateCompletedStreams()
                     }
 
                     _streams.Remove(stream.StreamId);
-                    ReturnStream(stream);
+
+                    // We don't want to return a stream for reuse that could still be draining
+                    if (stream.DrainExpirationTicks >= now)
+                    {
+                        ReturnStream(stream);
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs]: If `stream.DrainExpirationTicks >= now` that means `stream.EndStreamReceived || stream.RstStreamReceived` in which case it should be perfectly fine to return the stream to the pool.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs [hunk 3]
```diff
@@ -1087,7 +1098,6 @@ private void MakeSpaceInDrainQueue()
                 }
 
                 _streams.Remove(stream.StreamId);
-                ReturnStream(stream);
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs]: This call to ReturnStream can also stay. While It's true that CompleteStream() also eventually causes a stream to be returned to a pool, that's a different stream. This is returning a stream that used to be in the _completedStreams queue.

**Review by jkotalik — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs (hunk 2 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs]: Do you want that added as a comment?

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs (hunk 3 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs]: Hmmm, are you sure? You'd want to have the same check as up above: Has the stream received END_STREAM or RST_STREAM.

I'd rather be safe and say that if we're removing streams on an unexpected path we shouldn't try to reuse them. Better to be safe than sorry.

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs]: Re-added

**Review by halter73 — APPROVED** [review]

**Pull request merged**